e2e-tests/gef/cypress/e2e/facility/unissued-facilities/unissued-facilities-facility-end-date.spec.js

```diff
@@ -0,0 +1,160 @@
+
+        cy.clickBackLink();
+
+        cy.url().should('eq', relative(`/gef/application-details/${dealId}/unissued-facilities/${facilityOneId}/about`));
+      });
+
+      it('should display error messages', () => {
+        facilityEndDate.facilityEndDateDay().clear().type('abcd');
+        cy.clickContinueButton();
+
+        errorSummary();
+        facilityEndDate.facilityEndDateError();
+      });
+
+      it('should redirect user to the unissued facility page when clicking continue', () => {
+        facilityEndDate.facilityEndDateDay().clear().type(dateConstants.todayDay);
+        facilityEndDate.facilityEndDateMonth().clear().type(dateConstants.todayMonth);
+        facilityEndDate.facilityEndDateYear().clear().type(dateConstants.todayYear);
+
+        cy.clickContinueButton();
+
+        cy.url().should('eq', relative(`/gef/application-details/${dealId}/unissued-facilities`));
+      });
+
+      it('should redirect user to the unissued facility page when clicking save and return', () => {
+        facilityEndDate.facilityEndDateDay().clear().type(dateConstants.todayDay);
+        facilityEndDate.facilityEndDateMonth().clear().type(dateConstants.todayMonth);
+        facilityEndDate.facilityEndDateYear().clear().type(dateConstants.todayYear);
+
+        cy.clickSaveAndReturnButton();
+
+        cy.url().should('eq', relative(`/gef/application-details/${dealId}/unissued-facilities`));
+      });
+    });
+
+    describe('Visiting from application preview page', () => {
+      beforeEach(() => {
+        cy.visit(relative(`/gef/application-details/${dealId}`));
+        applicationPreview.facilitySummaryListTable(4).facilityEndDateAction().click();
+      });
+
+      it('should redirect user to the about unissued facility page when clicking back link', () => {
+        cy.url().should('eq', relative(`/gef/application-details/${dealId}/unissued-facilities/${facilityOneId}/facility-end-date/change`));
+
+        cy.clickBackLink();
+
+        cy.url().should('eq', relative(`/gef/application-details/${dealId}/unissued-facilities/${facilityOneId}/change`));
+      });
+
+      it('should display error messages', () => {
+        facilityEndDate.facilityEndDateDay().clear().type('abcd');
+        cy.clickContinueButton();
+
+        errorSummary();
+        facilityEndDate.facilityEndDateError();
+      });
+
+      it('should redirect user to the application details page when clicking continue', () => {
+        facilityEndDate.facilityEndDateDay().clear().type(dateConstants.todayDay);
+        facilityEndDate.facilityEndDateMonth().clear().type(dateConstants.todayMonth);
+        facilityEndDate.facilityEndDateYear().clear().type(dateConstants.todayYear);
+        cy.clickContinueButton();
+
+        cy.url().should('eq', relative(`/gef/application-details/${dealId}`));
+      });
+
+      it('should redirect user to the application details page when clicking save and return', () => {
```

Comment:
just to check, do we have any tests for testing data has been saved when coming back to the page? And also, what happens if data is entered which has validation errors and save and return is clicked?

Reply:
Theres a more detailed api test to cover saveAndReturn behaviour -  `gef-ui\api-tests\facility-end-date.post.api-test.js` & unit tests to cover validation in more detail `gef-ui\server\controllers\facility-end-date\validation.test.ts`.
I'll add an e2e test that the data is saved on returning

Reply:
thanks and it would be good to also have an e2e test to ensure that behaviour on save and return with validation errors occurs